### ISSUE TITLE
Ivan Smith rule to accommodate certain tax authorities for closed orders (to avoid unneeded automatic credits and debits)

### ISSUE THREAD

**Issue by kentbower** [kbower/tickettest1#2235]
_Imported from trac ticket 2235.
Created by **michelle**
Opened in trac: Mar 20, 2013, 8:02:37 AM
Last modified in trac: Aug 3, 2015, 11:14:11 AM_
***
_by michelle:_
See also #2219 

(The majority of these are being caused because Loft believes the Tax Authority needs to be 440 but it is finalized as something else.)

```

0318359SD58:
Closed "taxruleid" was updated from [802] to [440]

0318350SB46:
Closed "taxruleid" was updated from [401] to [440]

0318351RW69:
Closed "taxruleid" was updated from [482] to [440]

0316350RS15:
Closed "taxruleid" was updated from [401] to [440]

0316356RI62:
Closed "taxruleid" was updated from [460] to [440]

0316356RI46:
Closed "taxruleid" was updated from [460] to [440]

0315350QT86:
Closed "taxruleid" was updated from [401] to [440]

0315350QT16:
Closed "taxruleid" was updated from [401] to [440]

0313359QG14:
Closed "taxruleid" was updated from [802] to [440]

0313354QC38:
Closed "taxruleid" was updated from [156] to [440]

0313350PY81:
Closed "taxruleid" was updated from [401] to [440]

0312351PC37:
Closed "taxruleid" was updated from [482] to [440]

0311359OX48:
Closed "taxruleid" was updated from [802] to [440]

0311359OV79:
Closed "taxruleid" was updated from [802] to [440]

0311350OP89:
Closed "taxruleid" was updated from [401] to [440]

0309356OL43:
Closed "taxruleid" was updated from [460] to [440]

0308355NZ73:
Closed "taxruleid" was updated from [ATH] to [440]

0308350NX41:
Closed "taxruleid" was updated from [401] to [440]

0301350LC21:
Closed "taxruleid" was updated from [401] to [440]
```

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 8:03:27 AM:_

_From Michelle:_

```
STORE/TAX CODES

50:  	440 (Emory)
	401 (Greenville)

51:	440 (Emory)
	482 (Sulphur Springs)

54:	440 (Emory)
	156 (Gun Barrel)

55: 	440 (Emory)
	ATH (Athens)

56:	440 (Emory)
	460 (Paris)

59:	440 (Emory)
	802 (Palestine)

```

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 10:13:10 AM:_

# Explanation
Technically, the stores listed above by Michelle have two possible tax authorities, based on a contract Ivan Smith has with the government of the city of Emory, which states that deliveries or pickups which were purchased from Emory will be taxable by Emory (440), regardless of where possession is taken (delivery or pickup or where pickup is made) and even if only part of the inventory was sold from Emory (site 'TN').

In other words, if any of the goods sold came out of Emory ('TN' warehouse), even if by transfer, then Emory (440) has claim to sales tax.  If, however, none of the inventory was sold from the 'TN' warehouse, then city of the order's site should receive the sales tax.

But sales staff historically can't be trusted to choose the correct tax authority, so Ivan Smith has the data setup such that, while writing sales, only 440 is available for the stores listed above.  

After an order is closed, if research finds that none of the inventory for the sale was sold from site 'TN', then the tax authority is updated to the written site's city's tax rule (also regardless of where delivery is taken, according to TX law).  Currently, this is manually updated as GERS/Loft will not allow the tax authority to change after the sale has been closed.

After this tax authority (tax rule, more accurately) has been updated, if the order is subsequently saved in Loft (even if just for a comment), and part of the order is still open, Loft sees there is only one valid taxrule available for the closed order, e.i., 440, and therefore chooses 440 it when it figures tax.  

Loft then inspects differences and, noting that the taxrule has been updated to 440 on the closed order, decides it must create an automatic adjustment to record that fact.

# Solution
For already closed orders, we will write a rule for Ivan Smith that allows for either taxrule, as shown above, dependent upon order site.  Luckily, the tax _rate_ will still be the same, so no automatic adjustment will be required.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 12:23:32 PM:_

(Changeset [[3931]](../commit/f9856e60ce241abbe0253a9c8c0a510f7c6d477c))
* add ability to change 'check_for_rule' value for a @rule_property from startup rule, re #2235

**Comment by kentbower** [comment]
_**rneville** trac comment on Mar 20, 2013, 12:31:47 PM:_

(Changeset [[3932]](../commit/f7bbd62be2d2e5291afe12562863177ca13d3d46))
* prevent $0 autoadjustments for Ivan Smith's Emory tax situation (closes #2235)

**Comment by kentbower** [comment]
_**rneville** trac status change on Mar 20, 2013, 12:31:47 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 2:05:41 PM:_

Kent,
 
yes. that's pretty accurate. Texas law currently is "where delivery is taken" but way back years ago there was a clause that included furniture that the tax could go where the goods were delivered from. Mr Terry & the city of emory found that clause. The contract they had together is still active & grandfathered into the New Texas Law until they vote again this year to see if it gets extended.  Ivan Smith "Inherited" that contract with the purchase of Terry's Furniture. Its only for stores 50,51,52,54,55,56 & 59 that are serviced out of TN warehouse which is located in Emory, TX.  No other stores serviced out of TN can ever be included. Store 52 is in Emory so their tax code will always be 440. Thats why I did not send it to you.
 
The only problem I see is writing the rule for already closed orders. Since the rule will allow for either tax code for those stores, can it be for open & closed? It is easier to capture where the merchandise is originally coming from at point of sale. I work reports daily for the day before "written" which includes open & closed for that day & day before "finalized" business to catch any splits that may need to be changed. Plus it needs to be done as soon as possible incase there are any credits written against these sales. It takes less than 10 minutes to work daily. If I have to work off of closed orders only, I will have to research the merchandise more & it will take alot longer.
 
In other words, can the rule state that for those stores , whether open or closed, as long as they have either of the 2 tax codes assigned, for loft to ignore validation/updating of the tax code on open & ignore automatic adjustment on closed.
 
I hope that made sense. If not, please call me.
Michelle

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 20, 2013, 4:43:59 PM:_

Design changed.  Ivan Smith wants to allow changes for open order as well.  They will add the zipcodes back into `taxrulezipcodes` and instead, we will create a special ivan smith permission so the 7 stores listed above cannot have taxruleid _changed_ (or set to anything but '440' for new order) unless the user has permission.

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 20, 2013, 4:43:59 PM:_ **reopened**

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 9:56:57 AM:_

Michelle,
    Your rules already have a definition for sites that are serviced by TN that I thought I could use for the permission which will force taxrule of '440' unless you have the permission to change it from '440'.

In your rules, it is defined as TNSERVICED = ['42', '50', '51', '52', '53', '54', '55', '56', '59', 'TN'], but you didn't mention 'TN', '42', or '53'.

Is it safe to enforce this permission for 'TN', '42', and '53' as well as the others?

Also, to verify, the permission enforces '440' based on the SO_STORE_CD, not the PU_DEL_STORE_CD, correct?

Thanks,
Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 11:20:16 AM:_

Kent,
Do not include 42 & yes it is based on SO_STORE_CD.
Thanks
michelle 
___
_Edit: it is ok to enforce the permission for 'TN' and '53' _

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 12:36:39 PM:_

I just thought of another question: what about orders which you've already
set which are then split?  My guess is that if the order is split there is
a high probability that the tax code may need to be reinspected to see
what's going on and possibly changed.

So, for the new "A" or "B", etc. ticket if the one splitting does not have
the permission: should the taxcode be forced to the same as the original or
should it be forced to 440?

Kent

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 1:09:05 PM:_

Sorry, thought of another potential problem.  If an unrelated site (maybe
01) creates a CRM or MCR/MDB against such an order, the permission isn't
programmed to play a role there, since the written store is '01', so anyone
will be able to change that one's tax rule id...

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 2:07:30 PM:_

Replying to [kbower](comment:11):
> I just thought of another question: what about orders which you've already
> set which are then split?  My guess is that if the order is split there is
> a high probability that the tax code may need to be reinspected to see
> what's going on and possibly changed.
> 
> So, for the new "A" or "B", etc. ticket if the one splitting does not have
> the permission: should the taxcode be forced to the same as the original or
> should it be forced to 440?
> 
> Kent


I check all splits anyway on the finalized report so it can just stay as the original. I just dont want the store to be able to change it.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 2:08:29 PM:_

Replying to [kbower](comment:12):
> Sorry, thought of another potential problem.  If an unrelated site (maybe
> 01) creates a CRM or MCR/MDB against such an order, the permission isn't
> programmed to play a role there, since the written store is '01', so anyone
> will be able to change that one's tax rule id...
> 
> On Thu, Mar 21, 2013 at 12:32 PM, Kent Bower <kb@retailarchitects.com>wrote:


we try to enforce same store codes when creating crm,mcr & mdb's as the original store code but in case it does happen , its ok.

**Comment by kentbower** [comment]
_**kbower** trac comment on Mar 21, 2013, 3:08:07 PM:_

(Changeset [[3939]](../commit/4ae59d9897f493e165fa582ea07f7ecfd92e597e))
* Ivan Smith permission for being able to change taxruleid when accomodating Emory, TX tax contract (closes #2235)

**Comment by kentbower** [comment]
_**kbower** trac status change on Mar 21, 2013, 3:08:07 PM:_ **closed**

**Comment by kentbower** [comment]
_**kbower** trac comment on Aug 3, 2015, 11:14:11 AM:_

(Changeset [[6619]](../commit/67133f8738c0b23808f7e2db2ed5764452e7273a))
* fixes defaulting of Emory tax rule (see #2235), partly broken in [r6595](../commit/cf1292d8cd6b4c112add0cf08207843ea3509814) #isf
